Sparse CSR CPU: add `torch.addmm`

IvanYashchuk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62180
* __->__ #65606

This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc @nikitaved @pearu @cpuhrsch @IvanYashchuk

Differential Revision: [D32366236](https://our.internmc.facebook.com/intern/diff/D32366236)

IvanYashchuk committed:
Sparse CSR CPU: add `torch.addmm`

This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

cpuhrsch commented on aten/src/ATen/native/mkl/SparseBlasImpl.cpp:
```diff
@@ -20,6 +22,40 @@ namespace mkl {
+  * `tensor` - 2D strided Tensor.
+  * `row_major` - controls the memory layout.
+*/
+c10::MaybeOwned<Tensor> prepare_dense_matrix_for_mkl(
```
Why do you need this overload and the one above? Is it for n-dim Tensor support?

IvanYashchuk replied:
The one above returns either row or column-major matrix, it doesn't matter. This overload returns the matrix in a specified layout.

cpuhrsch commented on aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:
```diff
@@ -134,87 +134,98 @@ void s_addmm_out_sparse_dense_worker(int64_t nnz, int64_t dim_i, int64_t dim_j,
+
+#if !AT_MKL_ENABLED()
+    if (mat2.is_sparse_csr() && result.is_sparse_csr()) {
+      TORCH_CHECK(false, "Calling addmm on sparse CPU tensors requires compiling PyTorch with MKL. Please use PyTorch built with MKL.");
```
nit: I know this is using AT_MKL_ENABLED instead of AT_USE_MKL_SPARSE, but for this particular condition, shouldn't the error message be the same (i.e. it needs MKL *and* linux)?

IvanYashchuk replied:
Right, use `AT_USE_MKL_SPARSE` should probably be used here instead of `AT_MKL_ENABLED`.

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32366236).

IvanYashchuk committed:
Update on "Sparse CSR CPU: add `torch.addmm`"


This PR adds `torch.addmm(c, a, b, alpha=1.0, beta=0.0, out=out)` variant with `a, b, c, out` all being sparse CSR tensors on CPU.

cc nikitaved pearu cpuhrsch IvanYashchuk

Differential Revision: [D32366236](https://our.internmc.facebook.com/intern/diff/D32366236)

[ghstack-poisoned]

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32366236).
